### PR TITLE
Add momentum operator

### PR DESCRIPTION
**Pull request by sidgoyal78**
This PR adds the implementation of momentum operator.

In summary, we want to perform the update with a new velocity vector, such that,
```
 velocity = mu * velocity + grad
 param = param - learning_rate * velocity
```
(where mu is the momentum coefficient).

### CODE REVIEW

File: paddle/operators/momentum_op.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    param_out->mutable_data<T>(ctx.GetPlace());
+    velocity_out->mutable_data<T>(ctx.GetPlace());
+
+    float mu = ctx.Attr<float>("mu");
+
+    auto p = EigenVector<T>::Flatten(*ctx.Input<Tensor>("Param"));
+    auto g = EigenVector<T>::Flatten(*ctx.Input<Tensor>("Grad"));
+    auto v = EigenVector<T>::Flatten(*ctx.Input<Tensor>("Velocity"));
+    float lr = ctx.Input<Tensor>("LearningRate")->data<float>()[0];
```
Comment (reyoung) [inline paddle/operators/momentum_op.h]: That might be not good for GPU. If the `LearningRate` is in GPU memory, we cannot get `float` directly.

Comment (sidgoyal78) [reply paddle/operators/momentum_op.h]: Okay.. Thanks.

Comment (sidgoyal78) [reply paddle/operators/momentum_op.h]: Fixed as per #4598

File: paddle/operators/momentum_op.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename Place, typename T>
+class MomentumOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto param_out = ctx.Output<framework::Tensor>("ParamOut");
```
Comment (dzhwinter) [inline paddle/operators/momentum_op.h]: These parameters will be better if can be named with an `auto *` to indicate the real type.

**Comment by dzhwinter** [comment]
Thanks for this PR! @sidgoyal78 . Since our book chapters heavily depend on these optimizer operators, so merge this PR ASAP. We can leave the name style unified work in the future.

**Pull request merged**